### ISSUE TITLE
Render problems when Initial Visibility Collapsed

### ISSUE THREAD

**Issue by FWennerdahl** [cefsharp/CefSharp#1412]
Using CefSharp.MinimalExample.Wpf I'm having problems when binding the Visibility property to the LoadStateChanged event. If Visibility.Visible is set once the the event reports isLoading to be false, the browser does not render unless I also set a specific Height.
### My changes

In MainViewModel.cs I added an event handler for the ChromiumWebBrowser.LoadingStateChanged event that set the properties to bind to:

```
void webBrowser_LoadingStateChanged(object sender, LoadingStateChangedEventArgs e)
{
    if (e.IsLoading)
    {
        BrowserHeight = "Auto";
        BrowserVisibility = Visibility.Collapsed;
    }
    else
    {
        BrowserHeight = "500";
        BrowserVisibility = Visibility.Visible;
    }
}
```

In MainView.xaml I added bindings for Visibility and Height:

```
<cefSharp:ChromiumWebBrowser Grid.Row="0"
    Address="http://www.google.com"
    WebBrowser="{Binding WebBrowser, Mode=OneWayToSource}"
    Title="{Binding Title, Mode=TwoWay}" 
    Visibility="{Binding BrowserVisibility, Mode=OneWay}"/>
    <!--Height="{Binding BrowserHeight, Mode=OneWay}"/>-->
```
### Expected result

My goal was to make the browser collapsed until the LoadingStateChanged event reports that loading is done. Once loading is done, the browser should be set to Visible.
### Actual result

The event is triggered, the property updated but the browser does not render the page. However:
- If the Height binding is uncommented in MainView.xaml, the browser renders using the specified height.
- If the Height binding is uncommented but the Height is set to "Auto" when loading is done, the browser does not render. This is effectively the same thing as not binding Height as the default value for Height is Double.NaN which is equivalent to "Auto".
### Details and working example

I've created a working example to demonstrate this using CefShark.MinimalExample using CefSharp 43.0.0 (commit 12109da):
https://github.com/FWennerdahl/CefSharp.MinimalExample/tree/render-problems/43.0.0

I also tested this on 43.0.1 with the same changes (commit 2efb9bb):
https://github.com/FWennerdahl/CefSharp.MinimalExample/tree/render-problems/43.0.1

I've been running on Windows 10 Enterprise N. I could not see anything in the cef debug.log that differs between the working and not working case.
## Additional info

We are using CefSharp in a WPF application where we're displaying multiple browser windows simultaneously. We set the Height and Visibility of the browsers through bindings to only show them when they are done loading. The Height is reported from Javascript ~100ms after loading is done. We've been able to gather the following:
1. When binding only Height, some of the the browsers occassionally do not render but take up space in the flow. This is random and feels like some sort of race condition. This is our main problem but we've been unable to reproduce this consistently.
2. When binding both Height and Visibility to loading state, no browser is rendered. However, when mousing over the browser space, the mouse cursor changes according to the element that should be present on the page, tooltips display and when clicking the browser it partially renders. This is consistenly reproduced.
3. When binding Visibility to loading state but setting a fixed Height we have been unable to reproduce any of the problems above.

I've also witnessed some strange behaviour in CefSharp.Wpf.Example which seems related to this. I applied the same Visibility binding as I did for CefShapr.MinimalExample to the cefsharp/43 branch:
https://github.com/FWennerdahl/CefSharp/tree/render-problems/43

The following steps demonstrates the problems:
1. Start the example. The home.html page does not display as expected.
2. Open a new tab. www.google.com is loaded properly.
3. Go back to the first tab. Navigate to www.google.com. The page does not display.
4. Hover the mouse over the middle of the page where the search text field should be located. The mouse pointer does not change.
5. Adjust the angle slightly with the "Tweak WPF rendering" controls. The page still does not show.
6. Hover the mouse across the middle of the page again. The mouse pointer changes to a caret.
7. Go to the second tab and back to the first. The page renders in the adjusted angle.

Please let me know if there is anything else I can fill in or if I can help debug this.

**Comment by amaitland** [comment]
@FWennerdahl Very detailed, thanks :+1: I don't have time to go through anything in detail at the moment as I'm just getting back from holidays.

I will make note that the actual underlying `Browser` is only created when the control has `Width/Height` dimensions, see the `CreateOffscreenBrowserWhenActualSizeChanged` method.

https://github.com/cefsharp/CefSharp/blob/cefsharp/43/CefSharp.Wpf/ChromiumWebBrowser.cs#L903

**Comment by amaitland** [comment]
Also rendering is stopped soon as the browser is `Invisible`. Pages should however continue to load.

https://github.com/cefsharp/CefSharp/blob/cefsharp/43/CefSharp.Wpf/ChromiumWebBrowser.cs#L908

**Comment by amaitland** [comment]
@FWennerdahl How'd you go with resolving your problem?

**Comment by FWennerdahl** [comment]
We've worked around it for now by explicitly setting an initial height of the browser element (point 3 under "Additional info" in my original post). This has the unfortunate effect of making the height expand visibly once the actual height of the contents is reported from Javascript.

**Comment by amaitland** [comment]
@FWennerdahl Do you have any plans on contributing a `PR`? Otherwise I'll close this.

**Comment by FWennerdahl** [comment]
At the moment I'm afraid I do not have the time to investigate this enough to be able to produce a PR. Since we have currently locked our usage to CefSharp 43 I do not know if this is still reproducable in newer releases. Were you able to reproduce it using my example @amaitland?

If we update our dependency to a newer version of CefSharp and this issue is still present I can create a new issue and reference this one, if you'd like?

**Comment by amaitland** [comment]
> Were you able to reproduce it using my example @amaitland?

I had a quick look. `CEF` doesn't like when you provide `0` values for `Width` and `Height`, the structure of the current code works around this by waiting until `ActualWidth` and `ActualHeight` have values. (`CreateOffscreenBrowserWhenActualSizeChanged` delays browser creation). 

Having the initial `Visibility.Collapsed` effectively reverts back to the old behavior, providing `0` values, which was known to cause exactly the problem your seeing. Regardless of the controls height you'd still need to give `CEF` some values (probably even `1x1` would keep it happy).

In version `47` you may be able to work around the problem, you can manually call `WasResized()`, just need to make sure the values are greater than `0`.

https://github.com/cefsharp/CefSharp/blob/cefsharp/47/CefSharp/IBrowserHost.cs#L210

> If we update our dependency to a newer version of CefSharp and this issue is still present I can create a new issue and reference this one, if you'd like?

Can keep this open for now.

**Comment by AndrewKorell** [comment]
I believe I've experienced something similar. I am loading the browser into a tabbed content pane, as well as a hide-able docking pane. In both cases the browser appears to freeze when the container size is changed. Everything works fine if I give the browser a fixed size.

When the browser freezes it looks like a snapshot of the browser. The control buttons appear to work. If I click my home button the status and progress bars show my home page loading as normal. 

I've tried several different iterations of calling browser.InvalidateVisual(), browser.UpdateLayout(), directly loading the width and length, and calling WasResized() when the containers OnSizeChanged event with no change in out come. 

My temporary solution is to simply dispose and reload the browser.

**Comment by amaitland** [comment]
> When the browser freezes it looks like a snapshot of the browser. The control buttons appear to work. If I click my home button the status and progress bars show my home page loading as normal.

Are you using `DPI > 100%`?

**Comment by AndrewKorell** [comment]
amaitland,
I don't believe I am. I am not setting Cef.EnableHighDPISupport() anywhere. I have also verified that the zoom level is also zero.

**Comment by amaitland** [comment]
> I have also verified that the zoom level is also zero.

`ZoomLevel` is unrelated, there's just a very specific `DPI` bug that stops rendering in `WPF`.

**Comment by amaitland** [comment]
For reference I have no plans to look into this further, by all means investigate further and submit a `PR`.

**Comment by AndrewKorell** [comment]
Yes, I see the issue: _High DPI stops redrawing on Resize #1571_. I am still investigating.

**Comment by Saravanan04** [comment]
When we will get this fix?

**Comment by amaitland** [comment]
In newer version you can take control over when the browser is created, just subclass `ChromiumWebBrowser` and call `CreateOffscreenBrowser` with a size greater than `0,0`. All the `IBrowserHost` methods for resizing/invalidating are publicly available.

Improvements in the form of a `PR` are welcome.
